### PR TITLE
Add GraphBinary serialization support to gremlin-javascript

### PR DESCRIPTION
**Pull request by ihoro**
https://issues.apache.org/jira/browse/TINKERPOP-2229

This is the first tiny step draft, there is a lot of chore work ahead. And this is the reason I would like to request comments in order to make sure I’m going the right direction and to apply comments/suggestions at the early stages.

Shall I continue with such structure?

### CODE REVIEW

**Comment by spmallette** [comment]
@ihoro  I just wanted to say thanks for getting started on this. I didn't get a chance to look at the code itself today, but perhaps we can get you some feedback for tomorrow. As an administrative point, I think it would be great if you could target the `3.5-dev` branch for this rather than `master` as i think it would be excellent to have this feature on the 3.5.x line.

**Review by jorgebay — COMMENTED** [review]
It's looking great, feel free to @ me for any follow up question.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js [hunk 1]
```diff
@@ -40,6 +42,7 @@ const responseStatusCode = {
 
 const defaultMimeType = 'application/vnd.gremlin-v3.0+json';
 const graphSON2MimeType = 'application/vnd.gremlin-v2.0+json';
+const graphBinaryMimeType = 'application/vnd.graphbinary-v1.0';
```
Comment (divijvaidya) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/driver/connection.js:45]: Note that graphBinary has two MimeTypes. `application/vnd.graphbinary-v1.0` and `application/vnd.graphbinary-v1.0-stringd`

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/graph-serializer.js [hunk 2]
```diff
@@ -98,6 +99,47 @@ class GraphSON2Writer {
   write(obj) {
     return JSON.stringify(this.adaptObject(obj));
   }
+
```
[the remark this answers is not in the document]
Comment (ihoro) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/graph-serializer.js:102]: Good idea, please check #1549

**Comment by ihoro** [comment]
FYI: by now I’ve implemented the minimal set of types to get a simple request like `g.V().values()` work with default gremlin-server.

**Comment by ihoro** [comment]
But I have issue with TraverserSerializer, according to the spec I expect `{bulk}` as Int (32 bits), but gremlin-server responds with Long/Int64. Could you please guide me here?
(this is the same question posted via discord)

**Comment by jorgebay** [comment]
Nice progress!!

It looks like the [GraphBinary documentation](https://github.com/apache/tinkerpop/blob/master/docs/src/dev/io/graphbinary.asciidoc#traverser) is incorrect about traverser, it's a long: https://github.com/apache/tinkerpop/blob/master/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/Traverser.java#L109

Example serializer: https://github.com/apache/tinkerpop/blob/master/gremlin-dotnet/src/Gremlin.Net/Structure/IO/GraphBinary/Types/TraverserSerializer.cs#L52

**Comment by ihoro** [comment]
> It looks like the [GraphBinary documentation](https://github.com/apache/tinkerpop/blob/master/docs/src/dev/io/graphbinary.asciidoc#traverser) is incorrect about traverser, it's a long

Please, consider #1554.

**Comment by ihoro** [comment]
I’ve found an issue with T type. I’ve implemented its serialization according to the [spec](https://tinkerpop.apache.org/docs/current/dev/io/#_t_3). What means I send it like `{type_code}{type_info}` followed by non-fully-qualified string, i.e. `{length}{text_value}`. For example, T.label is sent as `0x2000 00000005 6c6162656c`. But gremlin-server does not seem to expect it with the following stack trace:
```
[WARN] DefaultChannelPipeline - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex(92) + length(1388) exceeds writerIndex(178): PooledUnsafeDirectByteBuf(ridx: 92, widx: 178, cap: 211)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:98)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.websocketx.Utf8FrameValidator.channelRead(Utf8FrameValidator.java:82)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IndexOutOfBoundsException: readerIndex(92) + length(1388) exceeds writerIndex(178): PooledUnsafeDirectByteBuf(ridx: 92, widx: 178, cap: 211)
        at io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1442)
        at io.netty.buffer.AbstractByteBuf.checkReadableBytes(AbstractByteBuf.java:1428)
        at io.netty.buffer.AbstractByteBuf.readBytes(AbstractByteBuf.java:895)
        at io.netty.buffer.AbstractByteBuf.readBytes(AbstractByteBuf.java:903)
        at org.apache.tinkerpop.gremlin.driver.ser.NettyBuffer.readBytes(NettyBuffer.java:132)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.StringSerializer.readValue(StringSerializer.java:37)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.StringSerializer.readValue(StringSerializer.java:28)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.readValue(SimpleTypeSerializer.java:59)
        at org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader.readValue(GraphBinaryReader.java:77)
        at org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader.read(GraphBinaryReader.java:102)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.EnumSerializer.readValue(EnumSerializer.java:66)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.EnumSerializer.readValue(EnumSerializer.java:44)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.readValue(SimpleTypeSerializer.java:59)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.read(SimpleTypeSerializer.java:47)
        at org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader.read(GraphBinaryReader.java:106)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.ByteCodeSerializer.getInstructionArguments(ByteCodeSerializer.java:56)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.ByteCodeSerializer.readValue(ByteCodeSerializer.java:41)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.ByteCodeSerializer.readValue(ByteCodeSerializer.java:30)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.readValue(SimpleTypeSerializer.java:59)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.read(SimpleTypeSerializer.java:47)
        at org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader.read(GraphBinaryReader.java:106)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.MapSerializer.readValue(MapSerializer.java:42)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.MapSerializer.readValue(MapSerializer.java:31)
        at org.apache.tinkerpop.gremlin.structure.io.binary.types.SimpleTypeSerializer.readValue(SimpleTypeSerializer.java:59)
        at org.apache.tinkerpop.gremlin.structure.io.binary.GraphBinaryReader.readValue(GraphBinaryReader.java:77)
        at org.apache.tinkerpop.gremlin.driver.ser.binary.RequestMessageSerializer.readValue(RequestMessageSerializer.java:55)
        at org.apache.tinkerpop.gremlin.driver.ser.GraphBinaryMessageSerializerV1.deserializeRequest(GraphBinaryMessageSerializerV1.java:175)
        at org.apache.tinkerpop.gremlin.server.handler.WsGremlinBinaryRequestDecoder.decode(WsGremlinBinaryRequestDecoder.java:77)
        at org.apache.tinkerpop.gremlin.server.handler.WsGremlinBinaryRequestDecoder.decode(WsGremlinBinaryRequestDecoder.java:43)
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
        ... 39 more
```
I’ve tried a fully-qualified string instead and gremlin-server is happy with it, i.e. it expects a fully-qualified string.

Q1: Is it a typo within the spec? Usually it explicitly mentions when a fully-qualified form is required, and non-FQ is expected otherwise.
Q2: If A1 is true then, I guess, the same is applied to other Enum based types like Cardinality, Order, and so on, right?
Q3: By the way, when gremlin-server errors in case of a non-fully-qualified string then a client gets no response and waits until its own timeout. Maybe a client should get something like 'wrong request' response instead of keeping silence from the server side? But this is a side question, not related to GraphBinary implementation for gremlin-javascript. Does it need a JIRA ticket creation for future consideration?

Could you please guide me from gremlin-core Java side perspective? I will help with a PR to fix the spec, if it gets confirmed.

**Comment by jorgebay** [comment]
hm... I thought T values, as the rest of enums were encoded as:

`{type_code}{type_info}{value_flag}{value}`

- type_code: `0x20` (1 byte)
- type_info: no data (0 bytes)
- value_flag: `0` (1 byte) - as not null
- value: `0`, `0`, `0`, `0x5` (length), followed by the string `0x6c` `0x61` `0x62` `0x65` `0x6c`

but looking at the java (reference) and C# implementations it looks like it expects a FQ string in the value instead, as you mention. The doc is incorrect.

About the client timing out, it's a client issue, it should error out, we can file a ticket for it.

**Comment by ihoro** [comment]
> but looking at the java (reference) and C# implementations it looks like it expects a FQ string in the value instead, as you mention. The doc is incorrect.

okay, thanks for confirmation, I will create respective PR then

**Comment by ihoro** [comment]
Please consider #1560 for docs correction.

**Comment by ihoro** [comment]
FYI (a teaser): one microservice of a project in production has got its integration tests (based on official gremlin-server) passed using current state of GraphBinary support implementation. This is not an evidence, just an interesting, and inspiring for me, fact.

**Comment by jorgebay** [comment]
> passed using current state of GraphBinary

NICE!!!

**Comment by ihoro** [comment]
I’ve been working on PathSerializer and I’ve found out that gremlin-server responds with fully-qualified `{labels}` and `{objects}`, i.e. as `0x0900...`. But the spec says `{labels} is a List` and `{objects} is a List`. You mentioned that Java based gremlin-server is the reference implementation, then, I guess, I should fix the docs respectively to make it explicit that FQ List is expected. Could you please confirm?

**Comment by ihoro** [comment]
> issue in the docs.

Thanks for the confirmation. Please, check respective PR #1566.

**Comment by spmallette** [comment]
`Binding` is a way to pass variables in Gremlin bytecode but it doesn't really have good applicability outside of situations where you end up using a lambda (which isn't often). It also has a weird implementation pattern that doesn't always work well in all languages - even in Java it uses awful `ThreadLocal` stuff to work properly. My personal feeling is that you can ignore it.

**Comment by ihoro** [comment]
Thank you for the confirmation.

And what about Graph type (0x10)? It has two names in the spec: Graph and TinkerGraph. And I also do not see related artifacts within gremlin-javascript. Probably, it’s some higher level concept from Java side what does not require direct (de)serialization in case of a driver like gremlin-javascript?

**Comment by spmallette** [comment]
you can't do the `Graph` type yet so you may skip that. we don't have a construct for that in javascript. hopefully that will be something we look at for 3.7.0.

**Comment by spmallette** [comment]
i'm not sure how others feel, but if this PR has reached the same level of compatibility as python (which doesn't cover all the serializers even as of today), I think it should be evaluated for merge for release of 3.5.3.

**Comment by ihoro** [comment]
Hi, I still have in plans to finish with some types like TraversalStrategy and BulkSet, as I see, according to gremlin-javascript and the tests, it’s important to implement them. Also, I have some technical concerns regarding numeric types (I was planning to discuss it with you).

My apologies, but my plans have been postponed due to my neighbour f(r)iend has been throwing rocks into my garden, breaking my house's windows, and trying to break through my doors.

If you want, please, continue according to your plans, i.e. merge and so on.

**Comment by spmallette** [comment]
@ihoro there are a lot of people that are shocked by your neighbor and are hoping he is made to stop very soon. take care.

**Comment by jorgebay** [comment]
@ihoro hope you and your family stay safe

**Comment by ihoro** [comment]
Hi,

I stuck with `should allow ReservedKeysVerificationStrategy` gremlin-javascript integration test for GraphBinary. Gremlin server complained that `ReservedKeysVerificationStrategy` Java class is not found.

Later I found that GraphSON sends Bytecode's `withStrategies` source without fully qualified class name, it simply sends `g:ReservedKeysVerificationStrategy` type based on JavaScript class name and, I guess (I did not research respective Java code), server side’s GraphSON implementation knows what Java class should be used. Thus, I got it that GraphBinary should send FQCN instead, and every `TraversalStrategy` of gremlin-javascript already provides respective `fqcn` field.

I got the next issue after that, Gremlin server was not erroring for reserved keys as the test expected. And I found that `ReservedKeysVerificationStrategy` class in gremlin-javascript has `fqcn=org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.EdgeLabelVerificationStrategy`. GraphSON does not use this field, that’s why the integration test is okay for it. I changed it to `org.apache.tinkerpop.gremlin.process.traversal.strategy.verification.ReservedKeysVerificationStrategy` and the integration test passed.

As long as this typo affects GraphBinary only, i.e. the new feature, I could integrate this typo fix within this PR. But, if you want to get it fixed as a separate PR then let me know, I’m going to keep it fixed localy meanwhile.

File: gremlin-javascript/src/main/javascript/gremlin-javascript/test/unit/graphbinary/AnySerializer-test.js [hunk 3]
```diff
@@ -520,6 +533,9 @@ describe('GraphBinary.AnySerializer', () => {
         ]
       },
 
+      // TRAVERSALSTRATEGY
+      // TODO: it's not expected to be deserialized, is it correct assumption?
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/test/unit/graphbinary/AnySerializer-test.js:552]: That's correct 👍

Comment (ihoro) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/test/unit/graphbinary/AnySerializer-test.js]: Thanks for the confirmation.

**Comment by ihoro** [comment]
I have got `integration` and `features` tests passed with current state of GraphBinary implementation (it's not 100% complete according to the spec). At least it works in my test env, it would be great to check it with official CI process. So, I would like to get the test process configured such way that both protocols are tested, GraphSON and GraphBinary. I've decided to make it so on the high level. From my quick overview, the lower levels seem to be tricky to configure. Please, consider [my approach](https://github.com/apache/tinkerpop/pull/1547/commits/6edeaee01254b5d095b3fb63cf68849627b6b2dd) and comment on, probably you would like to structure it another way according to the entire TinkerPop Maven project.

If you are okay with that env var introduced, quite global one, then... probably, it's better to name it more officially instead of `*_TEST_*`, and follow some naming agreement if any; because of it's part of production code, i.e. it can be used in a prod case as well, not only in test one.

What do you think?

**Comment by jorgebay** [comment]
I think we should would avoid using env variables for `Connection` properties default directly.

In order to facilitate testing, we could use the same env variable but use it to create the `Client` / `Connection` instance in the test modules. For example, in the `test/helper.js` module (pseudocode):

```javascript
exports.getClient = function getClient(traversalSource) {
  return new Client(serverUrl, { traversalSource, mimeType: process.env.YOUR_TEST_VAR });
};
```

That way modules within the `lib` don't have to be aware about test env vars.

**Comment by jorgebay** [comment]
> Could you please consider [this change](https://github.com/apache/tinkerpop/pull/1547/commits/bbbc4943e23078d3701b94d41964dccad35b0426).

Looks great 👍

**Comment by ihoro** [comment]
FYI: it was a rebase with conflicts resolved, going to align GraphBinary code with eslint rules added to the project

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/binary/internals/BulkSetSerializer.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        bulk = Number(bulk); // coersion to Number is required for BigInt if LongSerializerNg is used
+        const item = new Array(bulk).fill(value);
+        v = v.concat(item);
+      }
+
+      return { v, len };
+    } catch (e) {
+      throw this.ioc.utils.des_error({ serializer: this, args: arguments, cursor, msg: e.message });
```
Comment (jorgebay) [inline gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/binary/internals/BulkSetSerializer.js:118]: As an enhancement, I think we can still make use of the original stack trace by keeping the original error.

Maybe we can create a ticket for it?

Comment (ihoro) [reply gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/binary/internals/BulkSetSerializer.js]: Please, check [this change](https://github.com/apache/tinkerpop/pull/1547/commits/1b71d7ff8d2f685842e4fcec56ce643ebb41481f).

**Comment by spmallette** [comment]
thanks for sticking with this @ihoro - great job! 

VOTE +1 (whoever merges, please tack in changelog and perhaps an Upgrade Docs entry)

**Comment by ihoro** [comment]
From my side, I think it’s ready for merge, if you are okay with the open topics left as is for now. See [GraphBinary.js](https://github.com/ihoro/tinkerpop/blob/gremlin-javascript-graph-binary-support/gremlin-javascript/src/main/javascript/gremlin-javascript/lib/structure/io/binary/GraphBinary.js).
I agree that it’s a good idea to push it forward and create additional PRs on demand.

Sorry for getting it so long, I’ve got some force majeure, you know.
Thank you all for such a great support!

**Comment by jorgebay** [comment]
I'll merge it and do the downstream merges `3.6-dev` and `master`.

Following those changes, I'll create a small PR for the changelog/updatedocs and I'll ping u folks.

**Comment by jorgebay** [comment]
I usually advocate for squash & merge commit GH workflow but given that it's a large changeset (10K LOC) I think there's value in keeping the history.

**Comment by jorgebay** [comment]
I've merged it into 3.6-dev but it's failing when running cucumber tests with graphbinary: https://github.com/apache/tinkerpop/runs/6704524572?check_suite_focus=true#step:4:11048

```
env CLIENT_MIMETYPE='application/vnd.graphbinary-v1.0' ./node_modules/.bin/cucumber-js --require test/cucumber ../../../../../gremlin-test/features/
```

Any ideas @ihoro @spmallette ?

**Comment by spmallette** [comment]
I've fixed the problems on 3.6-dev/master - had to do with the new `Merge` enum and new testing of `GraphComputer` in GLVs. Both needed to be accounted for in the new serializer: e23a963c92bd6190a9b6b070d8dd4999a42d7471

> I usually advocate for squash & merge commit GH workflow but given that it's a large changeset (10K LOC) I think there's value in keeping the history.

agree with that choice in this case. glad you kept the history.

**Pull request merged**